server/src/main/java/org/opensearch/cluster/node/DiscoveryNode.java

```diff
@@ -567,10 +567,10 @@ private static Map<String, DiscoveryNodeRole> rolesToMap(final Stream<DiscoveryN
     private static Map<String, DiscoveryNodeRole> roleMap = rolesToMap(DiscoveryNodeRole.BUILT_IN_ROLES.stream());
 
     public static DiscoveryNodeRole getRoleFromRoleName(final String roleName) {
-        if (roleMap.containsKey(roleName) == false) {
-            throw new IllegalArgumentException("unknown role [" + roleName + "]");
+        if (roleMap.containsKey(roleName)) {
+            return roleMap.get(roleName);
         }
-        return roleMap.get(roleName);
+        return new DiscoveryNodeRole.UnknownRole(roleName, roleName, false);
```

Comment:
It seems we should not use `UnknownRole` here but introduce the concept of `DynamicRole`, complementing the builtin roles. Why: the `UnknownRole` is supposed to be used to deal with migrations when some of the builtin roles may be removed in new version but still present in the old one. The `DynamicRole` would help to make introduce such distinction.

Reply:
Thanks. Good  point. As we support dynamic/custom roles, it will be hard to differentiate `UnknownRole` (for example deprecated built-in role like `master`) and `DynamicRole` role (for example `ml` role) as both are not in built-in role list. So how about we just rename the `UnknownRole` as `DynamicRole`?